### ISSUE TITLE
Questions dealing with mbox files over 4GB 

### ISSUE THREAD

**Issue by cleidigh** [thundernest/import-export-tools-ng#452]
@Luclinuxer wrote

Good morning, first of all, thank you for the great application. I'm in need of help. I use the ImportExportTools NG 14.0.0 updated on October 4th (I use it on Debian, ImportExportTool installed on Thunderbird 115.3.1 (64 bits). After the routine system update, Thunderbir was updated and ImportExportTools NG 14.0.0 was different. I I was no longer able to export the entire structure of an email account, as I could before. So, I only exported the "incoming" folder. When I went to test whether the backup was working properly, by importing this "incoming" folder, the program warned me that the file was more than 4 gb and it didn't matter (something about java was written in the message). Can you help me with this? What to do, step by step? Thank you \o/

**Comment by cleidigh** [comment]
@Luclinuxer 
I am not sure how you ended up posting on an old commit so I created an issue for you here.

You are running into the 4GB limit which honestly I have not yet managed to completely deal with.

I didn't completely follow your upgrade comments, however, I think you arrived with Thunderbird v115.3.1 (latest) and ImportExportTools v14.0.0 which finally got released today.

Before we go into solving, please acknowledge that we are in the same place. I am deleting your other post as that is a separate problem.
@cleidigh

**Comment by Luclinuxer** [comment]
Exactly. I ask your forgiveness for the POST in an old location. So, I have Thunderbird v115.3.1 (64-bit) and ImportExportTools v14.0.0 running on a Debian. I've been using ImportExportTool for years without any problems. The problems I'm having are 2. The first: before, I right-clicked on the name of the email account I wanted to back up (incoming, sent, trash, etc...). And I clicked on export with subfolders, all the emails within them, to a directory chosen by me. I can't do it anymore, starting today. Right clicking opens the "export account" but nothing happens. I can only save by right clicking on "entry" of this account and export it all. Detail, I can't import files larger than 4 GB (this is the least problem, as I can try hard to delete as many emails as possible from now on).

**Comment by cleidigh** [comment]
@Luclinuxer 
Absolutely no apologies necessary, just wanted to both be in the same place!

So we are facing a couple of things here. The first is I partially messed up the export account, The container folder is not created and the top is exported as account.sbd 

If i understand you are saying there is nothing created?
I want to start here because we should be able to use these files.
@cleidigh

**Comment by Luclinuxer** [comment]
Dear @cleidigh 

I'll speak as a layman, not an expert, to help you even more: it's as if I gave the command to "export folder, with subdirectories" and it simply doesn't work, it doesn't work. And it was something simple, right? I clicked on the account, on the right side of Thunderbird (ex: lucas@gmail.com) and the option (among others) "export folder with its subdirectories" appeared. It does not work. What works: If I click, for example, on "input", or on "trash" and simply export. Tested. I exported. Then I deleted the content of the "input" and imported it. Ok (but one by one...."entrance", "trash bin"....). And that's when I discovered that files larger than 4 GB didn't matter (I was forced to delete an old part of the "input" content and save again. I managed not to lose everything because I had a backup from June 2023 (always saved the full Thunderbird identity/profile, just in case

**Comment by cleidigh** [comment]
@Luclinuxer 
Ok let's clarify a couple of things first (going backwards)

- at the end you talked about not losing anything because you have a backup.
Can you explain why you would lose something?

- What do you want to do, export the account for backup purposes?
- If you right click the account you should only see three options : Export Account, options and help.  Do you see the full menu, did it change?

@cleidigh

**Comment by Luclinuxer** [comment]
@cleidigh

- Perfect. I want to do what I always do. Backup of the email accounts I have (all of them), sent messages, incoming messages. And that's exactly it, when I click on the account name (her email) with the right button, I see: export account, options and help.
- I talked about losing emails because ImportTools is my email backup tool and once a month I backup all folders.

- Its functionality for me is exactly this: saving "incoming, sent, trash..." in my folder/directory.
The day I need to reinstall my Debian (or Slacl or Arch, when I use them) I go there and import everything correctly, without complications.

**Comment by cleidigh** [comment]
@Luclinuxer 
Thanks. 
So there are at least two issues we are dealing with that we have to work around.

The first is the Export account is partially messed up, however, we should see the main structure so we have to check that.

The second is the issue of folders larger than 4GB. This effects both imports and exports. Exports unfortunately are not cleanly delt with, it will will fail without warning.

To be safe as we deal with this you can do a profile export which will save everything including your large inbox even though it is over 4GB.

The first test I would like to do is the Export account. 

To get some information we want to use the Error console. To get the window  use Control-Shift-J
There is a trash can to clear the console. We will want to do that before tests. To copy the log you can use right click and copy all messages (to clipboard) you can then paste here.

So first test: 

- open console and clear
- have an empty directory for our export 
- right click on the account and choose the Export account 
- you should see the dialog for where to save the export, choose the directory you made
- go to the console, save all messages and post here
- go to the Export directory and tell me what is there

@cleidigh

**Comment by erickbonas** [comment]
Dear @cleidigh

I can't import mbox files larger than 4gb. How can I get around this problem?

My file is 20gb.

Thank you for your attention.

**Comment by Luclinuxer** [comment]
Dear @cleidigh

I copied these before cleaning as you asked

1696505700569	addons.xpi	WARN	Checking /usr/lib/thunderbird/distribution/extensions for addons
Found 0 public keys and 0 secret keys (0 protected, 0 unprotected) RNPLib.jsm:551:15
services.settings: Failed to load last_modified.json: TypeError: NetworkError when attempting to fetch resource. Utils.sys.mjs:322
1696505701646	addons.xpi	WARN	Addon with ID thunderbird-compact-light@mozilla.org already installed, older version will be disabled
1696505701646	addons.xpi	WARN	Addon with ID thunderbird-compact-dark@mozilla.org already installed, older version will be disabled
Esta página está em modo de compatibilidade (Quirks Mode). O layout da página pode ser afetado. Para usar o modo padrão (Standards Mode), inclua “<!DOCTYPE html>” no início. background.html
Missing resource in locale pt-BR: messenger/treeView.ftl
Missing resource in locale pt-BR: messenger/openpgp/openpgp.ftl
Trying to load /usr/lib/thunderbird/libotr.so OTRLib.sys.mjs:65:11
Trying to load libotr.so from system's standard library locations OTRLib.sys.mjs:65:11
Trying to load libotr.so.5 from system's standard library locations OTRLib.sys.mjs:65:11
Successfully loaded OTR library libotr.so.5 from system's standard library locations OTRLib.sys.mjs:73:13
NS_ERROR_FAILURE: Component returned failure code: 0x80004005 (NS_ERROR_FAILURE) [nsIAutoCompleteInput.popup] LoginManagerChild.sys.mjs:149
Alguns cookies estão usando incorretamente o atributo “SameSite“ 3
mailnews.smtp: 
error { target: TCPSocket, isTrusted: true, name: "NetworkTimeoutError", message: "Network", errorCode: 2152398862, srcElement: TCPSocket, eventPhase: 0, bubbles: false, cancelable: false, returnValue: true, … }
SmtpClient.jsm:442:17
Missing resource in locale pt-BR: devtools/client/toolbox.ftl

**Comment by Luclinuxer** [comment]
I don't know how to clean the console, I'm here watching, trying

**Comment by Luclinuxer** [comment]
Dear @cleidigh

I did it, I cleaned the console. I clicked on the account and export account, saved it in a folder on the desktop and copied the console messages below.
The folder, directory in which I saved it is empty

NotFoundError: No such JSWindowActor 'DevToolsFrame' 5
tb.ui.interaction.message_display - The key length must be limited to 72 characters.
Uncaught (in promise) Error: Component returned failure code: 0x80550006 [nsIMsgFolder.messages]
    buildAndExportMbox chrome://mboximport/content/mboximport/modules/mboxImportExport.js:383
    exportFoldersToMbox chrome://mboximport/content/mboximport/modules/mboxImportExport.js:331
    exportAccount chrome://mboximport/content/mboximport/mboximport.js:1005
mboxImportExport.js:383
Missing resource in locale pt-BR: devtools/client/toolbox.ftl

**Comment by cleidigh** [comment]
@Luclinuxer 
That was very helpful!
As best I can tell it's a Thunderbird error starting to export the first folder. I am assuming this is your large mbox. I think we have to copy maybe two haves to LocalFolders and then export from there.
@cleidigh

**Comment by erickbonas** [comment]
@cleidigh 

This mbox file is really big!

I performed this import in March, I was using Ubuntu 20.04 at the time,  and I had no issues. However, yesterday, when I was using Windows 10 to import this file again, I encountered this problem.  

I believe there was an update during this time period, as I had previously been able to import these files on Windows without any issues.

I'm using this mbox for a labor lawsuit. It contains emails dating back to 2015! I have these files in Thunderbird in Ubuntu, but I need them in Windows. 
 
Thank you!

**Comment by cleidigh** [comment]
@erickbonas 
Ok I am not totally clear on the history, however, if you need to keep this folder as one individual mbox then we can manually copy it in and hopefully Thunderbird will deal with it.
Give me a few hours as I have a meeting and I will give you the steps.
@cleidigh

**Comment by erickbonas** [comment]
Ok , no worries, thank you for your help!

**Comment by Luclinuxer** [comment]
@cleidigh

If I only click on the "inbox" of the account I'm trying to export in its entirety, and the "inbox" is no more than 4 GB, everything is fine, I can do it. In other words, do I have to do this manual backup, first the "inbox", then the "sent" folder?

**Comment by cleidigh** [comment]
@erickbonas 
Ok here is the manual method to get your megabox %-} into Thunderbird 
I cannot guarantee anything, but I think it should work.

- create a new folder for import in Local Folders, I will refer to as F
- on the folder, right click and choose IETNG \ Open folder directory 
- you should see the top level folders including F with no extension and a size of zero
- now close Thunderbird 
- copy your mbox into the directory, note it has to have the same name as F
- windows will warn on overwrite, say ok
- restart Thunderbird 
- the folder F will look empty
- right click on F and choose Folder properties 
- within the dialog click Repair folder
- you should see Rebuilding summary in the status bar.
- exit the dialog and you should see your messages 

Let me know your results
@cleidigh

**Comment by cleidigh** [comment]
@Luclinuxer 
Yes you can export the folders as individual mbox files and you should be good.
@cleidigh

**Comment by erickbonas** [comment]
@cleidigh 

Finally works!

Thank you so much for your help!

**Comment by Luclinuxer** [comment]
@cleidigh

Thank you for your help and clarification. Maybe, in the future, with a new version of Thunderbird, I will be able to backup my accounts like I did before (it was a little faster, it can still be done). Grateful

@Luclinuxer

**Comment by cleidigh** [comment]
@Luclinuxer 
Absolutely, fixing the account export will happen in the first maintenance release. I will also let you know when I have beta. Keep in mind the new method does error fixups so it will be slower.
@cleidigh

**Comment by cleidigh** [comment]
@Luclinuxer 
All

I have a beta that will import and export mboxes over 4GB up to 30GB.
There is still a very rare edge condition that can munge messages together that I will address in b2.

I also fixed the account export.

You can grab here:
- #466 
@cleidigh

**Comment by Luclinuxer** [comment]
@cleidigh
Your import program is excellent. I await any corrections, changes, ok? Thanks

**Comment by scratchwave** [comment]
> @erickbonas Ok here is the manual method to get your megabox %-} into Thunderbird I cannot guarantee anything, but I think it should work.
> 
> * create a new folder for import in Local Folders, I will refer to as F
> * on the folder, right click and choose IETNG \ Open folder directory
> * you should see the top level folders including F with no extension and a size of zero
> * now close Thunderbird
> * copy your mbox into the directory, note it has to have the same name as F
> * windows will warn on overwrite, say ok
> * restart Thunderbird
> * the folder F will look empty
> * right click on F and choose Folder properties
> * within the dialog click Repair folder
> * you should see Rebuilding summary in the status bar.
> * exit the dialog and you should see your messages
> 
> Let me know your results @cleidigh

I attempted to do this in macOS for two mbox files that were larger than 4GB, in separate attempts, and the 'repair' did not work as expected. That is to say, the repair probably did work as intended, by recreating the empty folders that the mbox files were meant to replace, however, mbox files remained alongside the repaired directories.

When copying the mbox files into their respective directories (with Thunderbird closed) the filesystem didn't throw a warning regarding overwriting the top level folder(s) of the same name and so I had to manually delete the original folder(s) created by Thunderbird, then copy the mbox file(s) across and then rename those to the original folders than had been deleted. This might be a macOS issue as the mbox files are not seen as folders but as a document.

I was able to import mbox files **under** 4GB using IETNG without issue. The mbox files were exported from Google Vault. The previous version of IETNG in a previous version of Thunderbird was able to import mbox files larger than 4GB.

It's also worth mentioning that the local folders are located on a network share. This is because we sometimes have to set up multiple local folders for data processing exercises, often with 200GB+ of mail exports, per local folder.

Just a further note to say this is such a great tool - thanks for the time spent on it! (If you can allow export of messages to PDF across all sub-directories of a folder that would be even more awesome!).

Let me know if there is anything else I can share with you.

Thank you!

Original error:

```
[JavaScript Application]
Mbox <NAME>.mbox larger than 4GB, skipping
```

**ENVIRONMENT**
macOS: 13.3
Thunderbird: 115.3.2 b(64-bit)
IETNG: v14.0.0

**Comment by cleidigh** [comment]
@scratchwave 
Thanks for your post and kind words!
Also thanks for your patience with these mbox issues, this is a major change.
Prior versions of IETNG basically did a blind copy of the mbox. This would work with mbox files over 4GB, however, there was no processing in particular From escaping. Good mboxes were fine, but mbox files without From escaping would have munged messages.

My new mbox code does proper From escaping, however, the file api was limited to 4GB, hence the warning. Bizarrely, just after release of v14.0.0 I randomly discovered the api had been updated to allow what I assume is effectively file system limits, even the documentation still says 4GB!!!
So I was thrilled at not having to come up with weird games.
So I already have a beta with > 4GB imports.
You can track and download from my 14.0.1 release status here : 
- #466 

Importantly there is a tricky edge case I am working on where a From line crosses a buffer boundary and can cause a message to be munged.
So you can try importing your large mbox, but there may be a couple of messages messed up.
Hopefully I can finish my fix in a few days, probably in b3 since I have a b2 ready.
Your testing will be a big help!

As for a structured PDF output, it's the first enhancement on my list as soon as I get things stable. I was hoping to have it in the first release, but Thunderbird v115 update was just too much.

Let me know if you try 14.0.1 
@cleidigh

**Comment by cleidigh** [comment]
@Luclinuxer 
I think I have account export, especially for pop3, back to normal.
Please let me know.
Grab b2 here: 
- #466 
@cleidigh

**Comment by cleidigh** [comment]
@Luclinuxer 
Great, glad you are back in business.

Yes because IETNG does file work it requires privileged permissions.
@cleidigh

**Comment by cleidigh** [comment]
All

I have an important fix in the mbox import code.
Basically it deals with From_ escaping across buffer boundaries.
With large imports there could be a few messages munged together.
If people can try some large imports that would be helpful. It's in b3: 
- #466 

Thanks 
@cleidigh

**Comment by techinmay** [comment]
If you’re dealing with an MBOX file larger than 4 GB, check whether

1. Duplicate emails are contributing to its size, especially messages with large attachments that have been stored multiple times.
2. Removing unnecessary duplicates can reduce the MBOX size and make it easier to manage, open, archive, or migrate.

In such cases, **[SysTools Thunderbird Duplicate Emails Remover](https://www.systoolsgroup.com/how-to/remove-duplicate-emails-from-thunderbird/)** can help scan MBOX files for duplicate emails and remove redundant copies. It supports duplicate removal both **within a single MBOX file and across multiple MBOX files**, making it useful when duplicates have accumulated through repeated backups or migrations.
